RubberduckTests/Rewriter/RewriteManagerTests.cs

```diff
@@ -0,0 +1,228 @@
+            {
+                mockSession.Verify(m => m.Invalidate(), Times.Once);
+            }
+        }
+
+        [Test]
+        [Category("Rewriter")]
+        public void CallingTheRewritingAllowedCallbackFromAnActiveAttributesSessionCallsInvalidateOnAllActiveSessions()
+        {
+            var rewritingManager = RewritingManager(out var mockFactory);
+            rewritingManager.CheckOutCodePaneSession();
+            var attributesSession =  rewritingManager.CheckOutAttributesSession();
+            rewritingManager.CheckOutCodePaneSession();
+            rewritingManager.CheckOutAttributesSession();
+
+            attributesSession.Rewrite();
+
+            foreach (var mockSession in mockFactory.RequestedCodePaneSessions().Concat(mockFactory.RequestedAttributesSessions()))
+            {
+                mockSession.Verify(m => m.Invalidate(), Times.Once);
+            }
+        }
+
+        [Test]
+        [Category("Rewriter")]
+        public void CallingTheRewritingAllowedCallbackFromANoLongerActiveCodePaneSessionDoesNotCallInvalidateOnAnyActiveSession_InactiveDueToRewrite()
+        {
+            var rewritingManager = RewritingManager(out var mockFactory);
+            var codePaneSession = rewritingManager.CheckOutCodePaneSession();
+            var attributesSession = rewritingManager.CheckOutAttributesSession();
+
+            rewritingManager.InvalidateAllSessions();
+
+            rewritingManager.CheckOutCodePaneSession();
+            rewritingManager.CheckOutAttributesSession();
+
+            codePaneSession.Rewrite();
+
+            foreach (var mockSession in mockFactory.RequestedCodePaneSessions()
+                .Concat(mockFactory.RequestedAttributesSessions())
+                .Where(session => session.Object != codePaneSession && session.Object != attributesSession))
+            {
+                mockSession.Verify(m => m.Invalidate(), Times.Never);
+            }
+        }
+
+        [Test]
+        [Category("Rewriter")]
+        public void CallingTheRewritingAllowedCallbackFromANoLongerActiveAttributesSessionDoesNotCallInvalidateOnAnyActiveSession_InactiveDueToRewrite()
+        {
+            var rewritingManager = RewritingManager(out var mockFactory);
+            var codePaneSession = rewritingManager.CheckOutCodePaneSession();
+            var attributesSession = rewritingManager.CheckOutAttributesSession();
+
+            rewritingManager.InvalidateAllSessions();
+
+            rewritingManager.CheckOutCodePaneSession();
+            rewritingManager.CheckOutAttributesSession();
+
+            codePaneSession.Rewrite();
+
+            foreach (var mockSession in mockFactory.RequestedCodePaneSessions()
+                .Concat(mockFactory.RequestedAttributesSessions())
+                .Where(session => session.Object != codePaneSession && session.Object != attributesSession))
+            {
+                mockSession.Verify(m => m.Invalidate(), Times.Never);
```

Comment:
Shouldn't the `Rewrite()` call should have invalidated the last two sessions created? I assumed that once rewritten, the sessions would be no longer valid.

Reply:
The point is that `Rewrite` gets called on a session that has been invalidated before. The test tests that it does not have an effect in the new sessions.